src/array_api_typing/_array.py

```diff
@@ -0,0 +1,70 @@
+__all__ = (
+    "Array",
+    "HasArrayNamespace",
+)
+
+from types import ModuleType
+from typing import Literal, Protocol
+from typing_extensions import TypeVar
+
+NamespaceT_co = TypeVar("NamespaceT_co", covariant=True, default=ModuleType)
+DTypeT_co = TypeVar("DTypeT_co", covariant=True)
+
+
+class HasArrayNamespace(Protocol[NamespaceT_co]):
+    """Protocol for classes that have an `__array_namespace__` method.
+
+    This `Protocol` is intended for use in static typing to ensure that an
+    object has an `__array_namespace__` method that returns a namespace for
+    array operations. This `Protocol` should not be used at runtime, for type
+    checking or as a base class.
+
+    Example:
+    >>> import array_api_typing as xpt
+    >>>
+    >>> class MyArray:
+    ...     def __array_namespace__(self):
+    ...         return object()
+    >>>
+    >>> x = MyArray()
+    >>> def has_array_namespace(x: xpt.HasArrayNamespace) -> bool:
+    ...     return hasattr(x, "__array_namespace__")
+    >>> has_array_namespace(x)
+    True
+
+    """
+
+    def __array_namespace__(
+        self, /, *, api_version: Literal["2021.12"] | None = None
+    ) -> NamespaceT_co: ...
+
+
+class HasDType(Protocol[DTypeT_co]):
+    """Protocol for array classes that have a data type attribute."""
+
+    @property
+    def dtype(self) -> DTypeT_co:
```

Comment:
In some obscure situations this can matter

```suggestion
    def dtype(self, /) -> DTypeT_co:
```

Reply:
SGTM.